Add rust

tbateson opened the pull request.

tbateson commented on src/main.ps1:
```diff
@@ -65,9 +65,15 @@ function Invoke-PwrScript($pkg) {
 	& $pkg
 	Invoke-PwrInit
 	Install-PwrPackage
+	if ($LASTEXITCODE -ne 0) {
+		throw "install package completed with exit code $LASTEXITCODE"
+	}
```
I added this code because the exit code from robocopy in `Install-PwrPackage` was failing the pipeline. It was not showing the failure in the pipeline until after `Test-PwrPackageInstall` completed, and the whole thing was quite disorienting. Remove if you think it doesn't make sense.

troyaf reviewed (changes requested):
Given your issues with robocopy, I tested with a commit that set the home variables to install directly into our package folder. Seems to work just as well. Let me know if you want to keep this, otherwise I would be okay with reverting.

tbateson commented on src/main.ps1:
```diff
@@ -65,9 +69,17 @@ function Invoke-PwrScript($pkg) {
+		if ($LASTEXITCODE -ne 0) {
+			throw "test package completed with exit code $LASTEXITCODE"
+		}
+		Invoke-DockerBuild 'image'
+		docker history image --no-trunc
```
```suggestion
```

tbateson commented on src/pkgs/rust.ps1:
```diff
@@ -0,0 +1,88 @@
+	if ($LASTEXITCODE -ne 0) {
+		throw "rustup default $($latest.Version) exit code $LASTEXITCODE"
+	}
+	foreach ($target in @('x86_64-pc-windows-msvc', 'x86_64-pc-windows-gnu', 'i686-pc-windows-msvc', 'i686-pc-windows-gnu')) {
```
```suggestion
	foreach ($target in @('x86_64-pc-windows-msvc')) {
```

tbateson commented on src/pkgs/rust.ps1:
```diff
@@ -0,0 +1,88 @@
+	}
+	$wantver = "rustc $($global:PwrPackageConfig.Version) "
+	Airpower exec 'file:///\pkg' {
+		Get-ChildItem $env:CARGO_HOME
+		Get-ChildItem $env:RUSTUP_HOME
```
```suggestion
```

tbateson commented on src/pkgs/vs-buildtools.ps1:
```diff
@@ -114,7 +114,7 @@ function global:Test-PwrPackageInstall {
+function global:Invoke-CustomDockerBuild($tag) {
 	Copy-Item Dockerfile -Destination "${env:ProgramFiles(x86)}\Dockerfile.vs-buildtools"
 	& docker build -f "${env:ProgramFiles(x86)}\Dockerfile.vs-buildtools" -t $tag "${env:ProgramFiles(x86)}\pkg"
 }
```
```suggestion
}

```

tbateson commented:
> Given your issues with robocopy, I tested with a commit that set the home variables to install directly into our package folder. Seems to work just as well. Let me know if you want to keep this, otherwise I would be okay with reverting.

Your changes look fine. I removed all of the targets except for the default 64-bit msvc target, as that's the rust-recommended one to use. If anyone wants gnu or 32-bit they can add it themself. This reduced the compressed file size from 330MB to 220MB.

I added a printout to display the gzipped size of the pkg directory to give a general ballpack size of the package. It's more useful for when we're adding new packages, if you know how to only run it for non-scheduled builds that would be the only time we need to see it. Otherwise I don't see significant performance issues with it running.
